File: problems/ruby/080721-pascals-triangle/code.rb
# https://leetcode.com/problems/pascals-triangle/
#
# @param {Integer} num_rows
# @return {Integer[][]}

def generate(num_rows)
  res = []
  prev_row = nil

  (1..num_rows).each do |i|
    cur_row = Array.new(i, 1)

    if prev_row
      (1...i - 1).each do |j|
        cur_row[j] = prev_row[j - 1] + prev_row[j]
      end
    end

    res.push(cur_row)
    prev_row = cur_row
  end

  res
end

puts generate(1).to_s
puts generate(2).to_s
puts generate(3).to_s
puts generate(4).to_s
puts generate(5).to_s

File: problems/javascript/07062024-1007210-pascals-triangle-2/code.js
/**
 * https://leetcode.com/problems/pascals-triangle-ii/
 *
 * @param {number} rowIndex
 * @return {number[]}
 */

const getRow = (rowIndex) => {
  const res = []
  let prevRow

  const numRows = rowIndex + 1
  for (let i = 1; i <= numRows; i++) {
    const curRow = new Array(i).fill(1)

    if (prevRow) {
      for (let j = 1; j < i - 1; j++) {
        curRow[j] = prevRow[j - 1] + prevRow[j]
      }
    }

    prevRow = curRow
    res.push(curRow)
  }

  return res[rowIndex]
}

console.log(getRow(3))
console.log(getRow(4))
console.log(getRow(5))
